ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,210 @@
+#[cfg(all(test, unit_test))]
+use {
+    futures::stream::{iter as stream_iter, TryStreamExt},
+    ipa_macros::Step,
+};
+
+#[cfg(all(test, unit_test))]
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, CustomArray, Field},
+    protocol::{
+        basics::Reveal, context::Context,
+        ipa_prf::boolean_ops::comparison_and_subtraction_sequential::compare_gt, RecordId,
+    },
+    secret_sharing::{replicated::semi_honest::AdditiveShare, WeakSharedValue},
+    seq_join::seq_join,
+};
+
+#[derive(Step)]
+#[cfg(all(test, unit_test))]
+pub(crate) enum Step {
+    Left,
+    Right,
+    Compare,
+    Reveal,
+}
+
+/// insecure quicksort using MPC comparisons and a key extraction function f
+///
+/// f takes as input an element in the slice and outputs the key by which we sort by
+/// follows partially the function signature of `sort_by_key`, see `https://doc.rust-lang.org/src/alloc/slice.rs.html#305-308`
+///
+/// set desc = true for descending ordering
+///
+/// This version of quicksort is insecure because it does not enforce the uniqueness of the sorted elements
+/// To see why this leaks information: assume that all elements are equal,
+/// then quicksort runs in time O(n^2) while for unique elements, it is only expected to run in time O(n log n)
+///
+/// The leakage can be fixed by appending a counter on each element that is unique to the element
+/// This adds another >= log N bits, where N is the amount of elements
+///
+/// This implementation of quicksort is in place and uses a stack instead of recursion
+/// It terminates once the stack is empty
+/// # Errors
+/// Will propagate errors from transport and a few typecasts
+#[cfg(all(test, unit_test))]
+pub async fn quicksort_by_key_insecure<C, K, F, S>(
+    ctx: C,
+    list: &mut [S],
+    desc: bool,
+    f: F,
+) -> Result<(), Error>
+where
+    C: Context,
+    S: Send + std::marker::Sync,
+    F: FnMut(&S) -> &AdditiveShare<K> + std::marker::Send + Clone + std::marker::Sync,
+    for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
+    K: WeakSharedValue + Field + CustomArray<Element = Boolean>,
+{
+    // create stack
+    let mut stack: Vec<(C, usize, usize)> = vec![];
```

Comment:
you can likely predict the depth of the circuit here, so `Vec::with_capacity(...)` should be better here

Reply:
the minimum length with be log length of input slice, in the worst case it will be length of the input slice (which is unlikely to happen). Is it still better to use vec::with_capacity even if I need to allocate exponentially more memory (log N vs N)? It probably depends on N which is the amount of events per users which we estimated to be at most 100k. so it would be 2*log_2(100k)< 34 vs 100k elements (notice that this stack implements depth first tree transgression)